llvm/include/llvm/Transforms/Utils/SYCLUtils.h

```diff
@@ -0,0 +1,137 @@
+#include "llvm/IR/GlobalVariable.h"
+#include "llvm/IR/Instructions.h"
+#include "llvm/IR/Operator.h"
+
+#include <functional>
+#include <string>
+#include <vector>
+
+namespace llvm {
+
+constexpr char ATTR_SYCL_MODULE_ID[] = "sycl-module-id";
+constexpr char ATTR_SYCL_OPTLEVEL[] = "sycl-optlevel";
+
+using CallGraphNodeAction = ::std::function<void(Function *)>;
+using CallGraphFunctionFilter =
+    std::function<bool(const Instruction *, const Function *)>;
+
+// Traverses call graph starting from given function up the call chain applying
+// given action to each function met on the way. If \c ErrorOnNonCallUse
+// parameter is true, then no functions' uses are allowed except calls.
+// Otherwise, any function where use of the current one happened is added to the
+// call graph as if the use was a call.
+// The 'functionFilter' parameter is a callback function that can be used to
+// control which functions will be added to a call graph.
+//
+// The callback is invoked whenever a function being traversed is used
+// by some instruction which is not a call to this instruction (e.g. storing
+// function pointer to memory) - the first parameter is the using instructions,
+// the second - the function being traversed. The parent function of the
+// instruction is added to the call graph depending on whether the callback
+// returns 'true' (added) or 'false' (not added).
+// Functions which are part of the visited set ('Visited' parameter) are not
+// traversed.
+
+void traverseCallgraphUp(
+    llvm::Function *F, CallGraphNodeAction NodeF,
+    SmallPtrSetImpl<Function *> &Visited, bool ErrorOnNonCallUse,
+    const CallGraphFunctionFilter &functionFilter =
+        [](const Instruction *, const Function *) { return true; });
+
+template <class CallGraphNodeActionF>
+void traverseCallgraphUp(
+    Function *F, CallGraphNodeActionF ActionF,
+    SmallPtrSetImpl<Function *> &Visited, bool ErrorOnNonCallUse,
+    const CallGraphFunctionFilter &functionFilter =
+        [](const Instruction *, const Function *) { return true; }) {
+  traverseCallgraphUp(F, CallGraphNodeAction(ActionF), Visited,
+                      ErrorOnNonCallUse, functionFilter);
+}
+
+template <class CallGraphNodeActionF>
+void traverseCallgraphUp(
+    Function *F, CallGraphNodeActionF ActionF, bool ErrorOnNonCallUse = true,
+    const CallGraphFunctionFilter &functionFilter =
+        [](const Instruction *, const Function *) { return true; }) {
+  SmallPtrSet<Function *, 32> Visited;
+  traverseCallgraphUp(F, CallGraphNodeAction(ActionF), Visited,
+                      ErrorOnNonCallUse, functionFilter);
+}
+
+/// Tells if this value is a bit cast or address space cast.
+bool isCast(const Value *V);
+
+/// Tells if this value is a GEP instructions with all zero indices.
+bool isZeroGEP(const Value *V);
+
+/// Climbs up the use-def chain of given value until a value which is not a
+/// bit cast or address space cast is met.
+const Value *stripCasts(const Value *V);
```

Comment:
What is the difference with [`Value::stripPointerCasts`](https://llvm.org/doxygen/classllvm_1_1Value.html#a08ad28e3c7d7c2f0c2ecb5ddcd1da355) and other similar methods?

Note that there is also [`llvm::getUnderlyingObject`](https://llvm.org/doxygen/namespacellvm.html#a3b1c5788b85ffa254be0e834edf5cf8f) and many more useful helpers in [`ValueTracking.h`](https://llvm.org/doxygen/ValueTracking_8h.html). We need to make sure not to reinvent them all

Reply:
I removed that from the current PR since these util functions are only used in ESIMD part.